### PR TITLE
Bring ha back

### PR DESCRIPTION
**Pull request by orangedeng**
Related norman PR https://github.com/rancher/norman/pull/167
Related types PR https://github.com/rancher/types/pull/459

### CODE REVIEW

**Comment by ibuildthecloud** [comment]
@orangedeng Can you explain the basic approach here?  What's the cattle instance, why a token, why another dialers, and why is the insecure flag set on the TLS connection.

**Comment by deniseschannon** [comment]
@ibuildthecloud This is from @orangedeng ```There are two parts for HA changes
1. We need to proxy the cluster connect request to leader instance. Refer to Implementation detail pkg/masterredirect/masterredirect.go.
2. We need to handle the API request which requires raw k8s api call. There are two way to do it.
    1. Proxy all the related APIs to master like we do for the cluster agent connect request. This is risky because we need to list all the APIs. It is very easy for developer to forget to add API in the list. So I don't use this solution.
    2. For now, we use cluster dialer to access k8s api. This dialer only works in leader rancher instance. So when the follower rancher instance is serving the API request and going to use cluster dialer, we can proxy use a proxy dialer instead. The proxy dialer will connect to leader rancher instance and talk to the real cluster dialer in leader instance. And the session manager will reuse this session for the same cluster request.

The proxy dialer implementation details in pkg/remotedialer/proxy_*. It is different from the current remote dialer and session. The current session is created by the cluster agent and used by the rancher server. Like following pattern:
1. Connect to create session: cluster agent -----> rancher server
2. Create dialer to use: rancher server ---- cluster agent(through session)----> Kubernentes cluster API
But in proxy dialer we will use this pattern:
1. Connect to create proxy session: follower rancher instance ------> leader rancher instance
2. Create dialer to use: follower rancher instance ------leader rancher instance(through proxy session)-------cluster agent(through leader session)------->Kubernentes cluster API
So I create a proxy dialer and proxy session to implement the logic above.```

**Comment by orangedeng** [comment]
1. The cattle instance means rancher pod in our deployment, inappropriate words.
2. why a token, I think you mean why I need a token to create a websocket. I think it will let the leader instance to validate the websocket request. It needs more security because we will accept request from that connection.
3. why another dialers, the new dialer will proxy all the exact k8s request to the leader instance. We won't need to worry about missing to set proxy for new API request.
4. and why is the insecure flag set on the TLS connection. I can fix this one.

File: app/app.go [hunk 1]
```diff
@@ -68,6 +81,10 @@ func buildScaledContext(ctx context.Context, kubeConfig rest.Config, cfg *Config
 }
 
 func Run(ctx context.Context, kubeConfig rest.Config, cfg *Config) error {
+	getID := func() string {
```
Comment (cjellick) [inline app/app.go]: whats the point of this wrapper function? only place it is called is on line 107 and getid(cfg) could have just been called directly

**Comment by orangedeng** [comment]
@cjellick add parameter `--advertise-address` to set the instance ip address. This is the unique ip address for each instance like 1.6. It will be used as the host ip address and port when proxying request to the leader rancher instance.

**Comment by alena1108** [comment]
Closing as approach to HA refactoring was changed since this PR was submitted, and the work will be covered by @ibuildthecloud @StrongMonkey PRs

**Pull request closed without merging**